code/modules/holodeck/computer.dm

```diff
@@ -110,207 +146,268 @@
 	. = TRUE
 	switch(action)
 		if("load_program")
-			var/program_to_load = text2path(params["type"])
-			if(!ispath(program_to_load))
-				return FALSE
-			var/valid = FALSE
+			var/program_to_load = params["id"]
+
 			var/list/checked = program_cache.Copy()
-			if(obj_flags & EMAGGED)
+			if (obj_flags & EMAGGED)
 				checked |= emag_programs
-			for(var/prog in checked)
-				var/list/P = prog
-				if(P["type"] == program_to_load)
+			var/valid = FALSE //dont tell security about this
+
+			for (var/prog in checked)//checks if program_to_load is any one of the loadable programs, if it isnt then it rejects it
+				var/list/check_list = prog
+				if (check_list["id"] == program_to_load)
 					valid = TRUE
 					break
-			if(!valid)
+			if (!valid)
 				return FALSE
-
-			var/area/A = locate(program_to_load) in GLOB.sortedAreas
-			if(A)
-				load_program(A)
+			//load the map_template that program_to_load represents
+			if(program_to_load)
+				load_program(program_to_load)
 		if("safety")
 			if((obj_flags & EMAGGED) && program)
 				emergency_shutdown()
-			nerf(obj_flags & EMAGGED)
+			nerf(obj_flags & EMAGGED,FALSE)
 			obj_flags ^= EMAGGED
-			say("Safeties restored. Restarting...")
+			say("Safeties reset. Restarting...")
+
+///this is what makes the holodeck not spawn anything on broken tiles (space and non engine plating / non holofloors)
+/datum/map_template/holodeck/update_blacklist(turf/placement, list/input_blacklist)
+	for (var/_turf in get_affected_turfs(placement))
+		var/turf/possible_blacklist = _turf
+		if (!istype(possible_blacklist, /turf/open/floor/holofloor))
+			if (istype(possible_blacklist, /turf/open/floor/engine))
+				continue
+			input_blacklist += possible_blacklist
```

Comment:
```suggestion
		if (istype(possible_blacklist, /turf/open/floor/holofloor))
			continue
		if (istype(possible_blacklist, /turf/open/floor/engine))
			continue
		input_blacklist += possible_blacklist
```
This could use a variable check instead of two istype ones.

Reply:
not sure what you mean by variable check but its formatted like the snippet above now

Reply:
Instead of checking the types of `possible_blacklist`, this property could be recorded on a variable that is checked here.
So instead of this:
```dm
if (istype(thing, /type_one) && istype(thing, /type_two))
	continue
```
You'd have something like this:
```dm
if (thing.has_property)
	continue
```